r/tests/testthat/test-Array.R

```diff
@@ -291,6 +291,17 @@ test_that("Timezone handling in Arrow roundtrip (ARROW-3543)", {
   expect_identical(read_feather(feather_file), df)
 })
 
+test_that("strptime", {
+  # array of strings
+  time_strings <- Array$create(c("2018-10-07 19:04:05", NA))
+  # array of timestamps (doesn't work if tz="" is added!)
```

Comment:
Why not?

Reply:
I haven't figured out yet how to pass `tz` argument to `strptime_arrow ` wrapper function as it is written now. After going through your comments, making the necessary changes then it will not be an issue I think.

Reply:
You're right that strptime in the C++ library doesn't take a timezone argument. Maybe it is expected that if there is a timezone, it will be encoded in the string and parseable by strptime (with the right format string)? But this gets us into the always tricky area of timezone-aware vs. timezone-naive data. @jorisvandenbossche do you have any thoughts/experience with this code?

Reply:
If I understand correctly [system strptime](https://man7.org/linux/man-pages/man3/strptime.3.html) is used so `%Z` or `%z` would work. E.g.: `2020-01-01 23:23:14 Europe/Amsterdam` would be captured by `format = "%Y-%m-%d %H:%M:%S %Z"`.
Capturing timezones would be great IMO but I would listen to Joris here for sure :).

Reply:
@rok do you know how the system `strptime` handles such a timezone if it is present? (the docs don't specify that, and the output struct doesn't have an entry for that information)

> Maybe it is expected that if there is a timezone, it will be encoded in the string and parseable by strptime (with the right format string)?

The problem here is that *if* a timezone is recorded in the Timestamp type's `tz` field, then the timestamp value is expected to be in UTC, and not localized to the timezone in question (which is what you get from just parsing the string without the timezone information). So basically that means the timestamp needs to be converted from the specific timezone to UTC (if `strptime` doesn't do that for us). And for now, that's not yet something we have implemented, I think (although at some point we probably should?)

Reply:
Right. I'm not sure, can you cast `timestamp[ms]` to `timestamp[ms, tz="UTC"]` or whatever (without modifying the values in the array, just to set the tz)?

Reply:
@jorisvandenbossche [it seems that](https://github.com/apache/arrow/blob/8e43f23dcc6a9e630516228f110c48b64d13cec6/cpp/src/arrow/util/value_parsing.h#L663) we don't really use or pass zone information even if strptime captures it. The following passes:
```
  options.timestamp_parsers = {TimestampParser::MakeStrptime("%Y-%m-%d %H:%M:%S %Z")};
  AssertConversion<TimestampType, int64_t>(type, {"1970-01-01 00:00:00 Etc/GMT+6,1970-01-01 00:00:00 UTC\n"}, {{0}, {0}}, options);
```
So timestamp's timezone is currently ignored and the local time is returned. It might be good to document this or even block `%z` and `%Z` to avoid surprises?

Reply:
> I'm not sure, can you cast `timestamp[ms]` to `timestamp[ms, tz="UTC"]` or whatever (without modifying the values in the array, just to set the tz)?

Casting actually works, but it's simply setting the tz and not changing the actual values, so it's not necessarily the behaviour you would expect (the behaviour would be correct if you assume the tz-naive data to be in UTC, but that seems a wrong assumption).

> So timestamp's timezone is currently ignored and the local time is returned. It might be good to document this or even block `%z` and `%Z` to avoid surprises?

Indeed, it seems we now simply ignore any timezone information in `strptime`:

```
>>> pc.strptime(["2012-01-01 01:02:03+01:00"], format="%Y-%m-%d %H:%M:%S%z", unit="s")
<pyarrow.lib.TimestampArray object at 0x7fad84855220>
[
  2012-01-01 01:02:03
]
>>> pc.strptime(["2012-01-01 01:02:03+01:00"], format="%Y-%m-%d %H:%M:%S%Z", unit="s").type
TimestampType(timestamp[s])
```

I can see some value in keeping that working, so you can at least parse strings that include such information (otherwise you would always get an error with arrow, or you would need to do some string preprocessing to be able to pass them to `strptime`). But then we certainly need to document that. 
On the other hand, if we want to support it in the future, that would change behaviour and erroring now might then be better ..

It seems that at least some `strptime` implementation support `%z` offsets, and store that in `tm->gmt_offset`, which we currently don't use (https://code.woboq.org/userspace/glibc/time/strptime_l.c.html#776). 
At least supporting fixed offsets (`%z`) seems doable (and the result could then be a timestamp type with `tz="UTC"`), properly supporting `%Z` timezone names will be harder.

Reply:
I've created [ARROW-12820](https://issues.apache.org/jira/browse/ARROW-12820) and referenced this discussion.
In context of this issue we could leave a reference to ARROW-12820 in the tests and postpone the timezone functionally?